File: windows/Studio.Core/Graph.cs
using System.Globalization;

namespace Lernheft.Studio;

/// <summary>
/// Rechnet Funktionsterme aus, die man eintippt: „x^2", „0,5x²-2x+1", „sin(x)".
/// Erlaubt sind + - * / ^, Klammern, sin cos tan sqrt abs ln log exp, pi und e.
/// Dieselbe Rechnung steckt in der iPad-App.
/// </summary>
public static class MathExpression
{
    public class ParseException(string message) : Exception(message);

    public static Func<double, double?> Compile(string text)
    {
        var cleaned = text.ToLowerInvariant();
        foreach (var (from, to) in new[]
                 {
                     ("f(x)=", ""), ("y=", ""), ("f(x) =", ""), ("y =", ""),
                     ("²", "^2"), ("³", "^3"), ("·", "*"), ("×", "*"), ("÷", "/"),
                     ("−", "-"), ("–", "-"), ("√", "sqrt"), ("π", "pi"), (",", ".")
                 })
        {
            cleaned = cleaned.Replace(from, to);
        }
        cleaned = cleaned.Replace(" ", "");
        if (cleaned.Length == 0) throw new ParseException("Tipp einen Funktionsterm ein, z. B. x^2 - 2x + 1");

        var parser = new Parser(Tokenize(cleaned));
        var node = parser.ParseExpression();
        if (!parser.AtEnd) throw new ParseException($"Das versteh ich nicht: „{parser.Rest}“");
        return x =>
        {
            var value = node.Value(x);
            return double.IsFinite(value) ? value : null;
        };
    }

    public static bool TryCompile(string text, out Func<double, double?>? function, out string? error)
    {
        try
        {
            function = Compile(text);
            error = null;
            return true;
        }
        catch (ParseException exception)
        {
            function = null;
            error = exception.Message;
            return false;
        }
    }

    private enum Kind { Number, Variable, Identifier, Symbol }

    private readonly record struct Token(Kind Kind, double Number = 0, string Name = "", char Symbol = '\0');

    private static List<Token> Tokenize(string text)
    {
        var tokens = new List<Token>();
        var index = 0;
        while (index < text.Length)
        {
            var character = text[index];
            if (char.IsDigit(character) || character == '.')
            {
                var start = index;
                while (index < text.Length && (char.IsDigit(text[index]) || text[index] == '.')) index++;
                var literal = text[start..index];
                if (!double.TryParse(literal, NumberStyles.Float, CultureInfo.InvariantCulture, out var value))
                    throw new ParseException($"Das versteh ich nicht: „{literal}“");
                tokens.Add(new Token(Kind.Number, value));
                continue;
            }
            if (char.IsLetter(character))
            {
                var start = index;
                while (index < text.Length && char.IsLetter(text[index])) index++;
                var name = text[start..index];
                // „2xsin(x)" oder „xx" soll wie 2·x·sin(x) rechnen.
                while (name.Length > 1 && name.StartsWith('x') && !IsKnown(name))
                {
                    tokens.Add(new Token(Kind.Variable));
                    name = name[1..];
                }
                tokens.Add(name == "x" ? new Token(Kind.Variable) : new Token(Kind.Identifier, Name: name));
                continue;
            }
            if ("+-*/^()".Contains(character))
            {
                tokens.Add(new Token(Kind.Symbol, Symbol: character));
                index++;
                continue;
            }
            throw new ParseException($"Das versteh ich nicht: „{character}“");
        }
        return tokens;
    }

    private static readonly string[] Functions = { "sin", "cos", "tan", "sqrt", "wurzel", "abs", "ln", "log", "exp" };

    private static bool IsKnown(string name) => name is "pi" or "e" || Functions.Contains(name);

    private abstract record Node
    {
        public abstract double Value(double x);
    }

    private sealed record Constant(double Number) : Node
    {
        public override double Value(double x) => Number;
    }

    private sealed record Variable : Node
    {
        public override double Value(double x) => x;
    }

    private sealed record Unary(string Name, Node Inner) : Node
    {
        public override double Value(double x)
        {
            var v = Inner.Value(x);
            return Name switch
            {
                "neg" => -v,
                "sin" => Math.Sin(v),
                "cos" => Math.Cos(v),
                "tan" => Math.Tan(v),
                "sqrt" or "wurzel" => v < 0 ? double.NaN : Math.Sqrt(v),
                "abs" => Math.Abs(v),
                "ln" => v <= 0 ? double.NaN : Math.Log(v),
                "log" => v <= 0 ? double.NaN : Math.Log10(v),
                "exp" => Math.Exp(v),
                _ => double.NaN
            };
        }
    }

    private sealed record Binary(char Operation, Node Left, Node Right) : Node
    {
        public override double Value(double x)
        {
            var a = Left.Value(x);
            var b = Right.Value(x);
            return Operation switch
            {
                '+' => a + b,
                '-' => a - b,
                '*' => a * b,
                '/' => b == 0 ? double.NaN : a / b,
                '^' => a < 0 && b != Math.Round(b) ? double.NaN : Math.Pow(a, b),
                _ => double.NaN
            };
        }
    }

    private class Parser(List<Token> tokens)
    {
        private int _index;

        public bool AtEnd => _index >= tokens.Count;

        public string Rest
        {
            get
            {
                if (AtEnd) return "";
                var token = tokens[_index];
                return token.Kind switch
                {
                    Kind.Number => token.Number.ToString(CultureInfo.InvariantCulture),
                    Kind.Variable => "x",
                    Kind.Identifier => token.Name,
                    _ => token.Symbol.ToString()
                };
            }
        }

        private Token? Peek() => AtEnd ? null : tokens[_index];

        private bool IsSymbol(char symbol) => Peek() is { Kind: Kind.Symbol } token && token.Symbol == symbol;

        public Node ParseExpression()
        {
            var node = ParseTerm();
            while (IsSymbol('+') || IsSymbol('-'))
            {
                var operation = tokens[_index++].Symbol;
                node = new Binary(operation, node, ParseTerm());
            }
            return node;
        }

        private Node ParseTerm()
        {
            var node = ParsePower();
            while (Peek() is { } token)
            {
                if (token.Kind == Kind.Symbol && token.Symbol is '*' or '/')
                {
                    _index++;
                    node = new Binary(token.Symbol, node, ParsePower());
                    continue;
                }
                // Verstecktes Malzeichen: 2x, 3(x+1), 2sin(x)
                var startsValue = token.Kind is Kind.Number or Kind.Variable or Kind.Identifier
                                  || (token.Kind == Kind.Symbol && token.Symbol == '(');
                if (!startsValue) break;
                node = new Binary('*', node, ParsePower());
            }
            return node;
        }

        private Node ParsePower()
        {
            var basis = ParseUnary();
            if (IsSymbol('^'))
            {
                _index++;
                return new Binary('^', basis, ParsePowerExponent());
            }
            return basis;
        }

        /// <summary>Im Exponenten darf ein Vorzeichen stehen: 2^-x.</summary>
        private Node ParsePowerExponent()
        {
            if (IsSymbol('-'))
            {
                _index++;
                return new Unary("neg", ParsePowerExponent());
            }
            return ParsePower();
        }

        private Node ParseUnary()
        {
            if (IsSymbol('-'))
            {
                _index++;
                // -x^2 ist -(x^2), wie in der Schule.
                return new Unary("neg", ParsePower());
            }
            if (IsSymbol('+'))
            {
                _index++;
                return ParsePower();
            }
            return ParsePrimary();
        }

        private Node ParsePrimary()
        {
            if (Peek() is not { } token) throw new ParseException("Der Term hört mittendrin auf.");
            switch (token.Kind)
            {
                case Kind.Number:
                    _index++;
                    return new Constant(token.Number);
                case Kind.Variable:
                    _index++;
                    return new Variable();
                case Kind.Identifier:
                    _index++;
                    if (token.Name == "pi") return new Constant(Math.PI);
                    if (token.Name == "e") return new Constant(Math.E);
                    if (!Functions.Contains(token.Name)) throw new ParseException($"Das versteh ich nicht: „{token.Name}“");
                    return new Unary(token.Name, ParsePrimary());
                case Kind.Symbol when token.Symbol == '(':
                    _index++;
                    var inner = ParseExpression();
                    if (!IsSymbol(')')) throw new ParseException("Da fehlt eine Klammer „)“.");
                    _index++;
                    return inner;
                default:
                    throw new ParseException($"Das versteh ich nicht: „{token.Symbol}“");
            }
        }
    }
}

/// <summary>
/// Baut aus einer Funktion echte Striche: Koordinatensystem mit Pfeilen, Skalenstriche,
/// Zahlen an den Achsen (als Textfelder) und den Graphen selbst.
/// </summary>
public static class GraphBuilder
{
    public record Settings(
        double XMin = -5, double XMax = 5, double YMin = -5, double YMax = 5,
        double Unit = 32, bool DrawAxes = true, bool DrawLabels = true,
        string Color = "#1A1F2B", double LineWidth = 2.4);

    public record Label(double X, double Y, double Width, string Text);

    public record Result(List<InkStroke> Strokes, List<Label> Labels, double Width, double Height);

    private const string AxisColor = "#1A1F2B";

    /// <summary><paramref name="originX"/>/<paramref name="originY"/> ist der Nullpunkt auf der Seite.</summary>
    public static Result Build(Func<double, double?>? function, Settings settings, double originX, double originY)
    {
        var strokes = new List<InkStroke>();
        var labels = new List<Label>();
        var unit = settings.Unit;
        var width = (settings.XMax - settings.XMin) * unit;
        var height = (settings.YMax - settings.YMin) * unit;

        (double X, double Y) Point(double x, double y) => (originX + x * unit, originY - y * unit);

        InkStroke Line(string color, double lineWidth, string kind, params (double X, double Y)[] points)
        {
            var stroke = new InkStroke { Id = InkIds.New(), Color = color, Width = lineWidth, Kind = kind };
            foreach (var (x, y) in points) stroke.Add(x, y, 0.8);
            return stroke;
        }

        if (settings.DrawAxes)
        {
            var left = Point(settings.XMin, 0);
            var right = Point(settings.XMax, 0);
            strokes.Add(Line(AxisColor, 1.8, "shape", left, right));
            strokes.Add(Line(AxisColor, 1.8, "shape", (right.X - 9, right.Y - 6), right, (right.X - 9, right.Y + 6)));

            var bottom = Point(0, settings.YMin);
            var top = Point(0, settings.YMax);
            strokes.Add(Line(AxisColor, 1.8, "shape", bottom, top));
            strokes.Add(Line(AxisColor, 1.8, "shape", (top.X - 6, top.Y + 9), top, (top.X + 6, top.Y + 9)));

            var step = Math.Max(1, (int)Math.Round(Math.Max(settings.XMax - settings.XMin, settings.YMax - settings.YMin) / 12));
            for (var value = (int)Math.Round(settings.XMin); value <= (int)Math.Round(settings.XMax); value += step)
            {
                if (value == 0) continue;
                var at = Point(value, 0);
                strokes.Add(Line(AxisColor, 1.5, "shape", (at.X, at.Y - 4), (at.X, at.Y + 4)));
                if (settings.DrawLabels) labels.Add(new Label(at.X - 10, at.Y + 5, 34, value.ToString(CultureInfo.InvariantCulture)));
            }
            for (var value = (int)Math.Round(settings.YMin); value <= (int)Math.Round(settings.YMax); value += step)
            {
                if (value == 0) continue;
                var at = Point(0, value);
                strokes.Add(Line(AxisColor, 1.5, "shape", (at.X - 4, at.Y), (at.X + 4, at.Y)));
                if (settings.DrawLabels) labels.Add(new Label(at.X - 32, at.Y - 11, 28, value.ToString(CultureInfo.InvariantCulture)));
            }
        }

        if (function is not null)
        {
            // In feinen Schritten abtasten; bei Lücken und Sprüngen neu ansetzen.
            var steps = Math.Max(2, (int)width);
            var segment = new InkStroke { Id = InkIds.New(), Color = settings.Color, Width = settings.LineWidth, Kind = "pen" };
            double? lastY = null;
            for (var index = 0; index <= steps; index++)
            {
                var x = settings.XMin + (settings.XMax - settings.XMin) * index / steps;
                var y = function(x);
                var inside = y is double value && value >= settings.YMin - 0.5 && value <= settings.YMax + 0.5;
                var jump = inside && lastY is double previous && Math.Abs(y!.Value - previous) > (settings.YMax - settings.YMin) * 0.6;
                if (!inside || jump)
                {
                    if (segment.Count >= 2) strokes.Add(segment);
                    segment = new InkStroke { Id = InkIds.New(), Color = settings.Color, Width = settings.LineWidth, Kind = "pen" };
                    lastY = inside ? y : null;
                    if (!inside) continue;
                }
                var (px, py) = Point(x, y!.Value);
                segment.Add(px, py, 0.9);
                lastY = y;
            }
            if (segment.Count >= 2) strokes.Add(segment);
        }
        return new Result(strokes, labels, width, height);
    }
}

File: ipad/LernheftPad/Graph.swift
import UIKit
import PencilKit

/// Rechnet Funktionsterme aus, die man mit der Tastatur eintippt: „x^2", „0,5x²-2x+1", „sin(x)".
/// Erlaubt sind + - * / ^, Klammern, sin cos tan sqrt abs ln log exp, pi und e.
enum MathExpression {
    enum ParseError: LocalizedError {
        case empty
        case unexpected(String)

        var errorDescription: String? {
            switch self {
            case .empty: return "Tipp einen Funktionsterm ein, z. B. x^2 - 2x + 1"
            case .unexpected(let part): return "Das versteh ich nicht: „\(part)“"
            }
        }
    }

    /// Macht aus dem eingetippten Text eine Funktion, die man beliebig auswerten kann.
    static func compile(_ text: String) throws -> (Double) -> Double? {
        var cleaned = text.lowercased()
        // Schreibweisen, die man auf dem iPad tippt oder aus dem Buch abschreibt
        for (from, to) in [
            ("f(x)=", ""), ("y=", ""), ("f(x) =", ""), ("y =", ""),
            ("²", "^2"), ("³", "^3"), ("·", "*"), ("×", "*"), ("÷", "/"),
            ("−", "-"), ("–", "-"), ("√", "sqrt"), ("π", "pi"), (",", ".")
        ] {
            cleaned = cleaned.replacingOccurrences(of: from, with: to)
        }
        cleaned = cleaned.replacingOccurrences(of: " ", with: "")
        guard !cleaned.isEmpty else { throw ParseError.empty }

        let tokens = try tokenize(cleaned)
        var parser = Parser(tokens: tokens)
        let node = try parser.parseExpression()
        guard parser.isAtEnd else { throw ParseError.unexpected(parser.rest) }
        return { x in
            let value = node.value(x)
            return value.isFinite ? value : nil
        }
    }

    // MARK: - Zerlegen

    private enum Token: Equatable {
        case number(Double)
        case variable
        case identifier(String)
        case symbol(Character)
    }

    private static func tokenize(_ text: String) throws -> [Token] {
        var tokens: [Token] = []
        let characters = Array(text)
        var index = 0
        while index < characters.count {
            let character = characters[index]
            if character.isNumber || character == "." {
                var literal = ""
                while index < characters.count, characters[index].isNumber || characters[index] == "." {
                    literal.append(characters[index])
                    index += 1
                }
                guard let value = Double(literal) else { throw ParseError.unexpected(literal) }
                tokens.append(.number(value))
                continue
            }
            if character.isLetter {
                var name = ""
                while index < characters.count, characters[index].isLetter {
                    name.append(characters[index])
                    index += 1
                }
                if name == "x" {
                    tokens.append(.variable)
                } else {
                    tokens.append(.identifier(name))
                }
                continue
            }
            if "+-*/^()".contains(character) {
                tokens.append(.symbol(character))
                index += 1
                continue
            }
            throw ParseError.unexpected(String(character))
        }
        return tokens
    }

    // MARK: - Aufbauen

    private indirect enum Node {
        case constant(Double)
        case variable
        case unary(String, Node)
        case binary(Character, Node, Node)

        func value(_ x: Double) -> Double {
            switch self {
            case .constant(let value):
                return value
            case .variable:
                return x
            case .unary(let name, let inner):
                let v = inner.value(x)
                switch name {
                case "neg": return -v
                case "sin": return sin(v)
                case "cos": return cos(v)
                case "tan": return tan(v)
                case "sqrt": return v < 0 ? Double.nan : sqrt(v)
                case "abs": return abs(v)
                case "ln": return v <= 0 ? Double.nan : log(v)
                case "log": return v <= 0 ? Double.nan : log10(v)
                case "exp": return exp(v)
                default: return Double.nan
                }
            case .binary(let operation, let lhs, let rhs):
                let a = lhs.value(x), b = rhs.value(x)
                switch operation {
                case "+": return a + b
                case "-": return a - b
                case "*": return a * b
                case "/": return b == 0 ? Double.nan : a / b
                case "^":
                    if a < 0, b != b.rounded() { return Double.nan }
                    return pow(a, b)
                default: return Double.nan
                }
            }
        }
    }

    private struct Parser {
        let tokens: [Token]
        var index = 0

        var isAtEnd: Bool { index >= tokens.count }

        var rest: String {
            guard !isAtEnd else { return "" }
            switch tokens[index] {
            case .number(let value): return "\(value)"
            case .variable: return "x"
            case .identifier(let name): return name
            case .symbol(let character): return String(character)
            }
        }

        private func peek() -> Token? { isAtEnd ? nil : tokens[index] }

        mutating func parseExpression() throws -> Node {
            var node = try parseTerm()
            while case .symbol(let operation)? = peek(), operation == "+" || operation == "-" {
                index += 1
                node = .binary(operation, node, try parseTerm())
            }
            return node
        }

        mutating func parseTerm() throws -> Node {
            var node = try parseUnary()
            while let token = peek() {
                if case .symbol(let operation) = token, operation == "*" || operation == "/" {
                    index += 1
                    node = .binary(operation, node, try parseUnary())
                    continue
                }
                // Verstecktes Malzeichen: 2x, 3(x+1), 2sin(x)
                let startsValue: Bool
                switch token {
                case .number, .variable, .identifier: startsValue = true
                case .symbol(let character): startsValue = character == "("
                }
                guard startsValue else { break }
                node = .binary("*", node, try parsePower())
            }
            return node
        }

        /// Hochzahlen binden stärker als das Minus davor: −x² ist −(x²), 2^−x geht auch.
        mutating func parsePower() throws -> Node {
            let base = try parsePrimary()
            if case .symbol("^")? = peek() {
                index += 1
                return .binary("^", base, try parseUnary())
            }
            return base
        }

        mutating func parseUnary() throws -> Node {
            if case .symbol("-")? = peek() {
                index += 1
                return .unary("neg", try parseUnary())
            }
            if case .symbol("+")? = peek() {
                index += 1
                return try parseUnary()
            }
            return try parsePower()
        }

        mutating func parsePrimary() throws -> Node {
            guard let token = peek() else { throw ParseError.empty }
            switch token {
            case .number(let value):
                index += 1
                return .constant(value)
            case .variable:
                index += 1
                return .variable
            case .identifier(let name):
                index += 1
                if name == "pi" { return .constant(Double.pi) }
                if name == "e" { return .constant(M_E) }
                let known = ["sin", "cos", "tan", "sqrt", "abs", "ln", "log", "exp"]
                guard known.contains(name) else { throw ParseError.unexpected(name) }
                let argument = try parsePrimary()
                return .unary(name, argument)
            case .symbol("("):
                index += 1
                let inner = try parseExpression()
                guard case .symbol(")")? = peek() else { throw ParseError.unexpected(")") }
                index += 1
                return inner
            case .symbol(let character):
                throw ParseError.unexpected(String(character))
            }
        }
    }
}

/// Achsenzahl eines Graphen – wird am Surface als kleines Textfeld gesetzt.
struct GraphLabel {
    var x: CGFloat
    var y: CGFloat
    var width: CGFloat
    var text: String
}

/// Baut aus einer Funktion echte Pencil-Striche: Koordinatensystem, Graph und Achsenzahlen.
enum GraphBuilder {
    struct Settings {
        var xMin: Double = -5
        var xMax: Double = 5
        var yMin: Double = -5
        var yMax: Double = 5
        /// Punkte pro Einheit – 32 entspricht genau einem Kästchen im karierten Papier.
        var unit: CGFloat = 32
        var drawAxes = true
        var drawLabels = true
        var color: UIColor = UIColor(hex: "#1A1F2B")
        var lineWidth: CGFloat = 2.4
    }

    struct Result {
        var strokes: [PKStroke] = []
        var labels: [GraphLabel] = []
        var size: CGSize = .zero
    }

    /// `origin` ist die Position des Nullpunkts auf der Seite (Inhaltskoordinaten).
    static func build(function: ((Double) -> Double?)?, settings: Settings, origin: CGPoint) -> Result {
        var result = Result()
        let unit = settings.unit
        let width = CGFloat(settings.xMax - settings.xMin) * unit
        let height = CGFloat(settings.yMax - settings.yMin) * unit
        result.size = CGSize(width: width, height: height)

        func point(_ x: Double, _ y: Double) -> CGPoint {
            CGPoint(x: origin.x + CGFloat(x) * unit, y: origin.y - CGFloat(y) * unit)
        }

        let axisColor = UIColor(hex: "#1A1F2B")
        if settings.drawAxes {
            // x-Achse mit Pfeilspitze
            if let stroke = InkBridge.line(
                points: [point(settings.xMin, 0), point(settings.xMax, 0)],
                width: 1.8,
                color: axisColor
            ) { result.strokes.append(stroke) }
            let xTip = point(settings.xMax, 0)
            for direction in [CGFloat(-1), CGFloat(1)] {
                if let stroke = InkBridge.line(
                    points: [CGPoint(x: xTip.x - 9, y: xTip.y + direction * 6), xTip],
                    width: 1.8,
                    color: axisColor
                ) { result.strokes.append(stroke) }
            }

            // y-Achse mit Pfeilspitze
            if let stroke = InkBridge.line(
                points: [point(0, settings.yMin), point(0, settings.yMax)],
                width: 1.8,
                color: axisColor
            ) { result.strokes.append(stroke) }
            let yTip = point(0, settings.yMax)
            for direction in [CGFloat(-1), CGFloat(1)] {
                if let stroke = InkBridge.line(
                    points: [CGPoint(x: yTip.x + direction * 6, y: yTip.y + 9), yTip],
                    width: 1.8,
                    color: axisColor
                ) { result.strokes.append(stroke) }
            }

            // Striche an den ganzen Zahlen
            let step = max(1, Int((max(settings.xMax - settings.xMin, settings.yMax - settings.yMin) / 12).rounded()))
            for value in stride(from: Int(settings.xMin.rounded()), through: Int(settings.xMax.rounded()), by: step) where value != 0 {
                let base = point(Double(value), 0)
                if let stroke = InkBridge.line(
                    points: [CGPoint(x: base.x, y: base.y - 4), CGPoint(x: base.x, y: base.y + 4)],
                    width: 1.5,
                    color: axisColor
                ) { result.strokes.append(stroke) }
                if settings.drawLabels {
                    result.labels.append(GraphLabel(x: base.x - 10, y: base.y + 6, width: 34, text: "\(value)"))
                }
            }
            for value in stride(from: Int(settings.yMin.rounded()), through: Int(settings.yMax.rounded()), by: step) where value != 0 {
                let base = point(0, Double(value))
                if let stroke = InkBridge.line(
                    points: [CGPoint(x: base.x - 4, y: base.y), CGPoint(x: base.x + 4, y: base.y)],
                    width: 1.5,
                    color: axisColor
                ) { result.strokes.append(stroke) }
                if settings.drawLabels {
                    result.labels.append(GraphLabel(x: base.x - 30, y: base.y - 9, width: 26, text: "\(value)"))
                }
            }
        }

        // Der Graph selbst: in Bildschirmschritten abtasten und bei Sprüngen unterbrechen
        if let function {
            var segment: [CGPoint] = []
            let steps = Int(width)
            for step in 0...max(1, steps) {
                let x = settings.xMin + (settings.xMax - settings.xMin) * Double(step) / Double(max(1, steps))
                if let y = function(x), y >= settings.yMin - 0.5, y <= settings.yMax + 0.5 {
                    segment.append(point(x, y))
                } else {
                    if segment.count >= 2, let stroke = InkBridge.line(points: segment, width: settings.lineWidth, color: settings.color) {
                        result.strokes.append(stroke)
                    }
                    segment = []
                }
            }
            if segment.count >= 2, let stroke = InkBridge.line(points: segment, width: settings.lineWidth, color: settings.color) {
                result.strokes.append(stroke)
            }
        }
        return result
    }

    /// Kurve wie ein Funktionsgraph: läuft durch alle Punkte, ohne Ecken und ohne Überschwingen.
    /// Nutzt ein monotones Spline (Fritsch–Carlson), wenn die Punkte von links nach rechts verlaufen.
    static func functionCurve(through points: [CGPoint], color: UIColor, width: CGFloat) -> PKStroke? {
        let sorted = points.sorted { $0.x < $1.x }
        guard sorted.count >= 2 else { return nil }
        // Zwei Punkte übereinander (gleiches x) verträgt ein Funktionsgraph nicht – dann weiche Kurve.
        for index in 1..<sorted.count where sorted[index].x - sorted[index - 1].x < 0.5 {
            return curve(through: points, smooth: true, color: color, width: width)
        }

        let count = sorted.count
        var slopes: [CGFloat] = []
        for index in 0..<(count - 1) {
            let dx = sorted[index + 1].x - sorted[index].x
            slopes.append((sorted[index + 1].y - sorted[index].y) / dx)
        }
        var tangents = [CGFloat](repeating: 0, count: count)
        tangents[0] = slopes[0]
        tangents[count - 1] = slopes[count - 2]
        for index in 1..<(count - 1) {
            if slopes[index - 1] * slopes[index] <= 0 {
                tangents[index] = 0                     // Hoch- oder Tiefpunkt: waagerechte Tangente
            } else {
                tangents[index] = (slopes[index - 1] + slopes[index]) / 2
            }
        }
        // Tangenten begrenzen, damit die Kurve nicht über die Punkte hinausschießt
        for index in 0..<(count - 1) where slopes[index] != 0 {
            let alpha = tangents[index] / slopes[index]
            let beta = tangents[index + 1] / slopes[index]
            let distance = hypot(alpha, beta)
            if distance > 3 {
                tangents[index] = 3 / distance * alpha * slopes[index]
                tangents[index + 1] = 3 / distance * beta * slopes[index]
            }
        }

        var curvePoints: [CGPoint] = []
        for index in 0..<(count - 1) {
            let p0 = sorted[index], p1 = sorted[index + 1]
            let dx = p1.x - p0.x
            let steps = max(6, Int(dx / 3))
            for step in 0..<steps {
                let t = CGFloat(step) / CGFloat(steps)
                let t2 = t * t, t3 = t2 * t
                let h00 = 2 * t3 - 3 * t2 + 1
                let h10 = t3 - 2 * t2 + t
                let h01 = -2 * t3 + 3 * t2
                let h11 = t3 - t2
                let y = h00 * p0.y + h10 * dx * tangents[index] + h01 * p1.y + h11 * dx * tangents[index + 1]
                curvePoints.append(CGPoint(x: p0.x + t * dx, y: y))
            }
        }
        curvePoints.append(sorted[count - 1])
        return InkBridge.line(points: curvePoints, width: width, color: color)
    }

    /// Glatte Kurve durch angetippte Punkte (Catmull-Rom) oder gerade Verbindungen.
    static func curve(through points: [CGPoint], smooth: Bool, color: UIColor, width: CGFloat) -> PKStroke? {
        guard points.count >= 2 else { return nil }
        guard smooth, points.count >= 3 else {
            return InkBridge.line(points: points, width: width, color: color)
        }
        var curvePoints: [CGPoint] = []
        let extended = [points[0]] + points + [points[points.count - 1]]
        for index in 1..<(extended.count - 2) {
            let p0 = extended[index - 1], p1 = extended[index]
            let p2 = extended[index + 1], p3 = extended[index + 2]
            let steps = max(6, Int(hypot(p2.x - p1.x, p2.y - p1.y) / 4))
            for step in 0..<steps {
                let t = CGFloat(step) / CGFloat(steps)
                let t2 = t * t, t3 = t2 * t
                let x = 0.5 * ((2 * p1.x) + (-p0.x + p2.x) * t + (2 * p0.x - 5 * p1.x + 4 * p2.x - p3.x) * t2 + (-p0.x + 3 * p1.x - 3 * p2.x + p3.x) * t3)
                let y = 0.5 * ((2 * p1.y) + (-p0.y + p2.y) * t + (2 * p0.y - 5 * p1.y + 4 * p2.y - p3.y) * t2 + (-p0.y + 3 * p1.y - 3 * p2.y + p3.y) * t3)
                curvePoints.append(CGPoint(x: x, y: y))
            }
        }
        curvePoints.append(points[points.count - 1])
        return InkBridge.line(points: curvePoints, width: width, color: color)
    }
}
